ql/src/java/org/apache/hadoop/hive/ql/optimizer/lineage/Generator.java

```diff
@@ -58,12 +67,63 @@
 public class Generator extends Transform {
 
   private static final Logger LOG = LoggerFactory.getLogger(Generator.class);
+  private static final String ALL = "ALL";
+  private static final String NONE = "NONE";
+  private static final Map<HiveOperation, Function<ParseContext, Boolean>> filterMap;
 
-  private final Set<String> hooks;
-  private static final String ATLAS_HOOK_CLASSNAME = "org.apache.atlas.hive.hook.HiveHook";
+  private final Predicate<ParseContext> statementFilter;
 
-  public Generator(Set<String> hooks) {
-    this.hooks = hooks;
+  static {
+    Map<HiveOperation, Function<ParseContext, Boolean>> map = new EnumMap<>(HiveOperation.class);
+    map.put(HiveOperation.CREATETABLE, parseContext -> parseContext.getCreateTable() != null);
+    map.put(HiveOperation.CREATETABLE_AS_SELECT, parseContext -> parseContext.getQueryProperties().isCTAS());
+    map.put(HiveOperation.CREATEVIEW, parseContext -> parseContext.getQueryProperties().isView());
+    map.put(HiveOperation.CREATE_MATERIALIZED_VIEW,
+        parseContext -> parseContext.getQueryProperties().isMaterializedView());
+    map.put(HiveOperation.LOAD,
+        parseContext -> !(parseContext.getLoadTableWork() == null || parseContext.getLoadTableWork().isEmpty()));
+    map.put(HiveOperation.QUERY, parseContext -> parseContext.getQueryProperties().isQuery());
+    filterMap = Collections.unmodifiableMap(map);
+  }
+
+  public static Generator fromConf(HiveConf conf) {
+    return new Generator(createFilterPredicateFromConf(conf));
+  }
+
+  static Predicate<ParseContext> createFilterPredicateFromConf(Configuration conf) {
+    Set<HiveOperation> operations = new HashSet<>();
+    boolean noneSpecified = false;
+    for (String valueText : conf.getStringCollection(HiveConf.ConfVars.HIVE_LINEAGE_STATEMENT_FILTER.varname)) {
+      if (ALL.equalsIgnoreCase(valueText)) {
+        return parseContext -> true;
+      }
+      if (NONE.equalsIgnoreCase(valueText)) {
+        noneSpecified = true;
```

Comment:
why not just check size of valueText end exit?

Reply:
ok, but instead of `NONE` option, can't we just check if it's Collections.emptySet()?

Reply:
If `Configuration.java.getStringCollection` returns empty set the loop does zero iterations, this `if` statement doesn't called and we end up with an always false lambda expression here:
```
return parseContext ->
        operations.stream().anyMatch(hiveOperation -> filterMap.get(hiveOperation).apply(parseContext));
```

Reply:
👍 do we need then NONE? can't we simply set `hive.lineage.statement.filter` to empty string?

Reply:
IMHO having a `NONE` setting is more user friendly but empty string is also allowed and means the same.
```
set hive.lineage.statement.filter=NONE;
```
vs
```
set hive.lineage.statement.filter=;
```

Reply:
exactly, why do we need NONE then
take a look at other configs like `TASK_THREADS_REMOTE_ONLY`, we supply the complete list
'ALL' is not declarative as well

Reply:
````hive.lineage.statement.filter=CREATETABLE, CREATETABLE_AS_SELECT, CREATEVIEW, CREATE_MATERIALIZED_VIEW, LOAD, QUERY````
i think it's more user-friendly and gives an understanding of what are the filters if one wants to disable some

Reply:
TASK_THREADS_REMOTE_ONLY has 6 values vs HiveOperation has ~40-50. Listing them all is not user friendly that is why I introduced the `ALL` value. `NONE` is the opposite of `ALL`.

Reply:
oh, sorry, I thought it was only above 6

Reply:
````
if (!pctx.getQueryProperties().isCTAS()
          && !pctx.getQueryProperties().isMaterializedView()
          && pctx.getQueryProperties().isQuery()
          && pctx.getCreateTable() == null
          && pctx.getCreateViewDesc() == null
          && (pctx.getLoadTableWork() == null || pctx.getLoadTableWork().isEmpty())) {
        LOG.debug("Not evaluating lineage");
        return pctx;
      }
````
@kasakrisz, we didn't log all operations before, why should we now?

Reply:
filterMap contains just a few operations, what happens in operation is not in a filter list?
````
filterMap.get(hiveOperation).apply(parseContext));
````
could there be NPE?

Reply:
> we didn't log all HiveOperation before, why should we now

Only if `org.apache.atlas.hive.hook.HiveHook` present:
https://github.com/apache/hive/blob/b775c708f7dd4cd2088e68f994460b42a22189a2/ql/src/java/org/apache/hadoop/hive/ql/optimizer/lineage/Generator.java#L75
Which never happens in any of our tests.

> filterMap contains just a few operations, what happens in operation is not in a filter list?

Good point. I found that we can not check every HiveOperation based on the parse context so I introduced a new enum only with the supported operation types.

Reply:
👍